File: packages/vis-agent/src/blockether/vis/activity.py
"""Immutable Activity receipts, counts, grouping and evidence.

For reading event data, see the
[SDK guide](https://vis.blockether.com/python-sdk.html#read-activity-and-view-receipts).
For authoring tool presentations, see the
[Activity presentation guide](https://vis.blockether.com/extension-api.html#activity-presentation).
"""

from __future__ import annotations

import json
from collections.abc import Mapping
from dataclasses import dataclass
from types import MappingProxyType
from typing import Any

from blockether.vis._contracts import definition, validate

from ._wire import freeze, to_wire


def _valid_handle_id(value: str) -> bool:
    limits = definition("activity", "handle_id")
    return (
        isinstance(value, str)
        and bool(value.strip())
        and len(value) <= limits["maxLength"]
        and len(value.encode("utf-8")) <= limits["x-vis-max-bytes"]
        and not any(ord(c) < 32 or ord(c) in (127, 8232, 8233) for c in value)
    )


@dataclass(frozen=True, slots=True)
class ActivityResource:
    """A typed reference to a resource involved in an invocation, identified by `id`."""

    type: str
    id: str


@dataclass(frozen=True, slots=True)
class ActivityDiffLine:
    """One classified diff line; `is_redacted` marks content withheld from the receipt."""

    kind: str
    text: str
    is_redacted: bool | None = None


@dataclass(frozen=True, slots=True)
class ActivityEvidence:
    """Text or structured change evidence attached to an invocation.

    Optional counts describe the observed changes; truncation and redaction flags
    explain why the displayed evidence may not contain the full content.
    """

    kind: str
    text: str
    lines: tuple[ActivityDiffLine, ...] | None = None
    additions: int | None = None
    deletions: int | None = None
    modifications: int | None = None
    is_truncated: bool | None = None
    is_redacted: bool | None = None


@dataclass(frozen=True, slots=True)
class ActivityRow:
    """One invocation's lifecycle, outcome and human-readable evidence.

    `id` and `sequence` preserve identity and ordering. `state` describes the
    outcome; `summary` is display text, not a replacement for status. Linked
    receipts appear as `children` with their shared `handle_id`. `presentation`
    contains the extension's selected content, while `evidence` retains
    engine-observed information.
    """

    id: str
    sequence: int
    operation: str
    presenter: str
    signal: str
    state: str
    summary: str
    resources: tuple[ActivityResource, ...]
    evidence: tuple[ActivityEvidence, ...]
    argument_key: str | None = None
    handle_id: str | None = None
    group_token: str | None = None
    duration_ms: int | None = None
    result_summary: str | None = None
    error_summary: str | None = None
    summary_format: str | None = None
    result_format: str | None = None
    is_truncated: bool | None = None
    children: tuple[ActivityRow, ...] | None = None
    presentation: Mapping[str, Any] | None = None

    @classmethod
    def _from_validated(cls, value):
        evidence = tuple(
            ActivityEvidence(
                **{
                    **item,
                    **(
                        {
                            "lines": tuple(
                                ActivityDiffLine(**line) for line in item["lines"]
                            )
                        }
                        if "lines" in item
                        else {}
                    ),
                }
            )
            for item in value["evidence"]
        )
        children = (
            {
                "children": tuple(
                    cls._from_validated(child) for child in value["children"]
                )
            }
            if "children" in value
            else {}
        )
        return cls(
            **{
                **freeze(value),
                "resources": tuple(ActivityResource(**r) for r in value["resources"]),
                "evidence": evidence,
                **children,
            }
        )


def _first_invocation_id(row: ActivityRow) -> str:
    return (
        row.children[0]
        if row.children and (row.handle_id is not None or row.operation == "shell")
        else row
    ).id


def _operation_group_label(operation: str, rows: list[ActivityRow]) -> str:
    labels = definition("activity", "row")["properties"]["operation"][
        "x-vis-group-labels"
    ]
    if operation in labels:
        return labels[operation]
    for row in rows:
        headline = (row.presentation or {}).get("headline", "").strip()
        if headline:
            return headline
    return operation


@dataclass(frozen=True, slots=True)
class ActivityArgumentGroup:
    """One operation with identical complete arguments; all invocation evidence is retained."""

    id: str
    rows: tuple[ActivityRow, ...]


def _argument_groups(
    rows: tuple[ActivityRow, ...],
) -> tuple[ActivityArgumentGroup, ...]:
    grouped: dict[tuple[str | None, str], list[ActivityRow]] = {}
    for row in sorted(rows, key=lambda row: row.sequence):
        key = (row.operation, row.argument_key) if row.argument_key else (None, row.id)
        grouped.setdefault(key, []).append(row)
    return tuple(
        ActivityArgumentGroup(_first_invocation_id(members[0]), tuple(members))
        for members in grouped.values()
    )


@dataclass(frozen=True, slots=True)
class ActivityGroup:
    """One exact operation across the block, ordered by first invocation.

    Shell polling remains child evidence. Unknown operations use their exact names.
    Disclosure state belongs to the reader, never to this receipt.
    """

    id: str
    label: str
    rows: tuple[ActivityRow, ...]

    @property
    def argument_groups(self) -> tuple[ActivityArgumentGroup, ...]:
        """Repeated arguments in first-entry order; unknown argument keys stay separate."""
        return _argument_groups(self.rows)


@dataclass(frozen=True, slots=True)
class ActivityCounts:
    """Invocation counts by lifecycle outcome, distinct from the visible row count."""

    running: int
    succeeded: int
    failed: int
    cancelled: int


@dataclass(frozen=True, slots=True)
class ActivityOmitted:
    """Rows omitted from this projection, with counts by omission classification."""

    rows: int
    by_classification: Mapping[str, int]


@dataclass(frozen=True, slots=True)
class ActivityProjection:
    """One form's Activity receipt, optionally a page of its durable history.

    Prefer `from_wire` to direct construction when accepting external data.
    `rows` and nested values are immutable; `to_wire` produces an independent
    mutable JSON-compatible copy. Grouping properties are computed, not stored.
    """

    state: str
    counts: ActivityCounts
    rows: tuple[ActivityRow, ...]
    omitted: ActivityOmitted
    history: Mapping[str, Any] | None = None

    @property
    def groups(self) -> tuple[ActivityGroup, ...]:
        """The same per-operation groups used by Companion and TUI; not serialized."""
        grouped: dict[str, list[ActivityRow]] = {}
        for row in sorted(self.rows, key=lambda row: row.sequence):
            grouped.setdefault(row.operation, []).append(row)
        return tuple(
            ActivityGroup(
                _first_invocation_id(rows[0]),
                _operation_group_label(operation, rows),
                tuple(rows),
            )
            for operation, rows in grouped.items()
        )

    @property
    def argument_groups(self) -> tuple[ActivityArgumentGroup, ...]:
        """Exact operation/argument pairs within this block; not serialized."""
        return _argument_groups(self.rows)

    @classmethod
    def from_wire(cls, value: Any) -> ActivityProjection:
        """Validate a receipt and construct immutable rows and nested records.

        Raises:
            ValueError: The schema, identity uniqueness, presentation bounds or
                history page limits are invalid.
        """
        validate("activity", "projection", value)
        ids = []
        leaf_count = 0

        def visit(rows):
            nonlocal leaf_count
            for row in rows:
                presentation = row.get("presentation")
                for carrier in (row, presentation):
                    if carrier is not None and "handle_id" in carrier:
                        if not _valid_handle_id(carrier["handle_id"]):
                            raise ValueError("invalid Activity handle id")
                sections = (
                    [presentation, *presentation.get("sections", [])]
                    if presentation is not None
                    else []
                )
                content = [
                    block for section in sections for block in section["content"]
                ]
                if any(
                    len(section[key].encode("utf-8")) > 512
                    for section in sections
                    for key in ("headline", "summary")
                ):
                    raise ValueError("activity headline or summary exceeds bound")
                for block in content:
                    if block["type"] == "progress" and "value" in block:
                        if not (0 <= block["value"] <= block["total"]):
                            raise ValueError("invalid activity progress")
                    if block["type"] == "table" and any(
                        len(cells) != len(block["columns"]) for cells in block["rows"]
                    ):
                        raise ValueError("invalid activity table width")
                ids.append(row["id"])
                children = row.get("children", [])
                if not children:
                    leaf_count += 1
                visit(children)

        visit(value["rows"])
        if len(set(ids)) != len(ids) or (
            "history" in value
            and (
                leaf_count
                > definition("activity", "projection")["properties"]["history"][
                    "x-vis-max-page-rows"
                ]
                or (
                    leaf_count > 1
                    and len(
                        json.dumps(
                            value, ensure_ascii=False, separators=(",", ":")
                        ).encode()
                    )
                    > definition("activity", "projection")["properties"]["history"][
                        "x-vis-page-target-bytes"
                    ]
                )
            )
        ):
            raise ValueError("invalid activity projection identity or byte bound")
        return cls(
            value["state"],
            ActivityCounts(**value["counts"]),
            tuple(ActivityRow._from_validated(row) for row in value["rows"]),
            ActivityOmitted(
                value["omitted"]["rows"],
                MappingProxyType(dict(value["omitted"]["by_classification"])),
            ),
            freeze(value["history"]) if "history" in value else None,
        )

    def to_wire(self) -> dict[str, Any]:
        """Return a fresh portable receipt without computed reader grouping fields."""
        return to_wire(self)

File: packages/vis-contract/src/com/blockether/vis/contract/activity.clj
(ns com.blockether.vis.contract.activity
  "Activity vocabulary, declarations and lossless history pages. Lifecycle stays in the engine."
  (:require [clojure.string :as str]
            [com.blockether.vis.contract.document :as document]
            [com.blockether.vis.contract.wire :as wire])
  (:import [java.nio.charset StandardCharsets]))

(def ^:private schema (document/schema-document "activity"))

(def presenters (set (map keyword (get-in schema ["$defs" "presenter" "enum"]))))

(def page-row-limit
  (get-in schema ["$defs" "projection" "properties" "history" "x-vis-max-page-rows"]))

(def page-byte-target
  (get-in schema ["$defs" "projection" "properties" "history" "x-vis-page-target-bytes"]))

(def resource-limit (get-in schema ["$defs" "row" "properties" "resources" "maxItems"]))

(def summary-byte-limit
  (get-in schema ["$defs" "section" "properties" "summary" "x-vis-max-bytes"]))

(def detail-byte-limit
  (get-in schema ["$defs" "evidence" "oneOf" 0 "properties" "text" "x-vis-truncate-bytes"]))

(def event-byte-limit (get-in schema ["$defs" "projection" "x-vis-max-event-bytes"]))

(def diff-line-byte-limit
  (get-in schema ["$defs" "diff_line" "properties" "text" "x-vis-truncate-bytes"]))

(defn valid-handle-id?
  "Opaque, bounded identity for receipts belonging to one live operation."
  [value]
  (and (document/valid-json? "activity" "handle_id" value)
       (<= (alength (.getBytes ^String value StandardCharsets/UTF_8))
           (long (get-in schema ["$defs" "handle_id" "x-vis-max-bytes"])))
       (not-any? #(or (< (int %) 32) (contains? #{127 8232 8233} (int %))) value)))

(defn valid-presentation?
  "Admit complete content and non-nested sections with brief headlines and summaries."
  [value]
  (let [value (wire/->wire value)]
    (and (document/valid-json? "activity" "presentation" value)
         (let [sections (cons value (get value "sections"))
               blocks (mapcat #(get % "content") sections)
               bytes #(alength (.getBytes ^String % StandardCharsets/UTF_8))]

           (and (or (not (contains? value "handle_id")) (valid-handle-id? (get value "handle_id")))
                (every? #(<= (long (bytes %)) (long summary-byte-limit))
                        (mapcat #(map % ["headline" "summary"]) sections))
                (every? (fn [block]
                          (case (get block "type")
                            "progress"
                            (or (nil? (get block "value"))
                                (<= (double (get block "value")) (double (get block "total"))))

                            "table"
                            (and (every? #(= (count %) (count (get block "columns")))
                                         (get block "rows"))
                                 (let [paths (get block "paths")]
                                   (or (nil? paths) (= (count paths) (count (get block "rows"))))))

                            true))
                        blocks))))))

(defn valid-projection?
  "Admit lossless projections; one indivisible invocation may exceed the page byte target."
  [value]
  (and (document/valid? "activity" "projection" value)
       (let [value
             (wire/->wire value)

             children
             #(get % "children")

             rows
             (mapcat #(tree-seq (comp seq children) children %) (get value "rows"))

             ids
             (map #(get % "id") rows)]

         (and (= (count ids) (count (set ids)))
              (every? #(or (not (contains? % "handle_id")) (valid-handle-id? (get % "handle_id")))
                      rows)
              (every? #(or (not (contains? % "presentation"))
                           (valid-presentation? (get % "presentation")))
                      rows)
              (or (nil? (get value "history"))
                  (and (<= (count (filter #(empty? (children %)) rows)) (long page-row-limit))
                       (or (<= (count (filter #(empty? (children %)) rows)) 1)
                           (<= (alength (.getBytes ^String (wire/json-str value)
                                                   StandardCharsets/UTF_8))
                               (long page-byte-target)))))))))

(defn from-wire
  "Valid projection in engine spelling, or nil. Never accepts retired owner/view keys."
  [value]
  (when (and (document/valid-json? "activity" "projection" value) (valid-projection? value))
    (wire/->engine value)))

(defn- first-invocation-id
  [row]
  (:id (if (and (seq (:children row)) (or (:handle-id row) (= "shell" (:operation row))))
         (first (:children row))
         row)))

(defn argument-groups
  "Group exact operation and argument-key pairs within one block. Missing keys stay
   separate; displayed summaries never prove equality. Preserve all invocation evidence."
  [rows]
  (let [ordered
        (sort-by :sequence rows)

        group-key
        (fn [row]
          (if-let [key (:argument-key row)]
            [:arguments (:operation row) key]
            [:row (:id row)]))

        grouped
        (group-by group-key ordered)]

    (mapv (fn [key]
            (let [members (get grouped key)]
              {:id (first-invocation-id (first members)) :rows members}))
          (distinct (map group-key ordered)))))

(defn operation-groups
  "One group per exact operation across the block, ordered by first entry.
   Known operations use canonical labels; otherwise use the first nonblank presentation
   headline in invocation order, falling back to the operation name when none is present.
   Preserve member order and shell evidence, with the first invocation as disclosure identity."
  [rows]
  (let [ordered
        (sort-by :sequence rows)

        grouped
        (group-by :operation ordered)]

    (mapv (fn [operation]
            (let [members
                  (get grouped operation)

                  row
                  (first members)]

              {:id (first-invocation-id row)
               :label (or (get-in schema
                                  ["$defs" "row" "properties" "operation" "x-vis-group-labels"
                                   operation])
                          (first (remove str/blank? (map (comp :headline :presentation) members)))
                          operation)
               :rows members}))
          (distinct (map :operation ordered)))))

(defn- content-copy-text
  [{:keys [type text columns rows label value total attachment-id]}]
  (case (name type)
    "table"
    (str/join "\n" (map #(str/join "\t" %) (cons columns rows)))

    "progress"
    (str label (when (some? value) (str ": " value "/" total)))

    (or text (str (name type) ": " label " (" attachment-id ")"))))

(defn- row-copy-text
  [{:keys [operation state duration-ms summary result-summary error-summary resources evidence
           presentation is-truncated children]} depth]
  (let [indent
        (apply str (repeat depth "  "))

        detail
        (fn [text]
          (when-not (str/blank? text) (map #(str indent "  " %) (str/split text #"\n" -1))))

        evidence-text
        (fn [{:keys [kind text lines is-truncated is-redacted]}]
          (str/join "\n"
                    (concat [(str (name kind) ":\n" text)]
                            (when (= "diff" (name kind))
                              (concat (map (fn [{:keys [kind text]}]
                                             (str (case (name kind)
                                                    "addition"
                                                    "+"

                                                    "deletion"
                                                    "-"

                                                    "context"
                                                    " "

                                                    "")
                                                  text))
                                           lines)
                                      (when is-truncated ["Diff truncated"])
                                      (when is-redacted ["Diff redacted"]))))))

        sections
        (when presentation (cons presentation (:sections presentation)))

        details
        (concat [summary (when result-summary (str "Result: " result-summary))
                 (when error-summary (str "Error: " error-summary))]
                (map #(str (name (:type %)) ": " (:id %)) resources)
                (map evidence-text evidence)
                (mapcat (fn [{:keys [headline summary content]}]
                          (concat [headline summary] (map content-copy-text content)))
                        sections)
                (when is-truncated ["Details truncated"]))]

    (str/join "\n"
              (concat [(str indent
                            operation
                            " [" (name state)
                            "]" (when (some? duration-ms) (str " (" duration-ms "ms)")))]
                      (mapcat detail details)
                      (mapcat #(vector "" (row-copy-text % (inc depth)))
                              (sort-by :sequence children))))))

(defn copy-text
  "Copy retained invocations from an engine-spelled projection, independent of grouping,
   disclosure and viewport. Include bounded evidence and omissions, never identity keys."
  [{:keys [rows omitted]}]
  (str/join "\n\n"
            (concat ["ACTIVITY"]
                    (map #(row-copy-text % 0) (sort-by :sequence rows))
                    (when (pos? (long (or (:rows omitted) 0)))
                      [(str (:rows omitted)
                            " step"
                            (when (not= 1 (:rows omitted)) "s")
                            " omitted · Activity limit")]))))
